# First Steps with the FXCM API

## Connecting to the API / Server

In [1]:
#importing libraries
import pandas as pd
import fxcmpy

In [5]:
# to start a connection between fxcm and jupyter Notebook
api = fxcmpy.fxcmpy(config_file= "fxcm.cfg")

In [6]:
# account details.
api.get_accounts()

t  ratePrecision accountId   balance  usdMr mc mcDate accountName  usdMr3  \
0  6              0   1682745  20181.15      0  N         1001682745       0   

  hedging  usableMargin3  usableMarginPerc  usableMargin3Perc    equity  \
0       Y       20181.15               100                100  20181.15   

   usableMargin   bus  dayPL  grossPL  
0      20181.15  1000      0        0

In [9]:
#account details.
api.get_account_ids()

[1682745]

In [8]:
#instruments fxcm provides for trading
api.get_instruments()

['EUR/USD',
 'USD/JPY',
 'GBP/USD',
 'USD/CHF',
 'EUR/CHF',
 'AUD/USD',
 'USD/CAD',
 'NZD/USD',
 'EUR/GBP',
 'EUR/JPY',
 'GBP/JPY',
 'CHF/JPY',
 'GBP/CHF',
 'EUR/AUD',
 'EUR/CAD',
 'AUD/CAD',
 'AUD/JPY',
 'CAD/JPY',
 'NZD/JPY',
 'GBP/CAD',
 'GBP/NZD',
 'GBP/AUD',
 'AUD/NZD',
 'USD/SEK',
 'EUR/SEK',
 'EUR/NOK',
 'USD/NOK',
 'USD/MXN',
 'AUD/CHF',
 'EUR/NZD',
 'USD/ZAR',
 'USD/HKD',
 'ZAR/JPY',
 'USD/TRY',
 'EUR/TRY',
 'NZD/CHF',
 'CAD/CHF',
 'NZD/CAD',
 'USD/HUF',
 'TRY/JPY',
 'USD/CNH',
 'AUS200',
 'ESP35',
 'FRA40',
 'GER30',
 'HKG33',
 'JPN225',
 'NAS100',
 'SPX500',
 'UK100',
 'US30',
 'Copper',
 'CHN50',
 'EUSTX50',
 'VOLX',
 'USDOLLAR',
 'US2000',
 'INDIA50',
 'USOil',
 'UKOil',
 'SOYF',
 'NGAS',
 'USOilSpot',
 'UKOilSpot',
 'WHEATF',
 'CORNF',
 'Bund',
 'XAU/USD',
 'XAG/USD',
 'EMBasket',
 'JPYBasket',
 'BTC/USD',
 'BCH/USD',
 'ETH/USD',
 'LTC/USD',
 'CryptoMajor',
 'EOS/USD',
 'XLM/USD',
 'ESPORTS',
 'BIOTECH',
 'CANNABIS',
 'FAANG',
 'CHN.TECH',
 'CHN.ECOMM',
 'USEquities',
 'A

In [10]:
instr = api.get_instruments()
instr

['EUR/USD',
 'USD/JPY',
 'GBP/USD',
 'USD/CHF',
 'EUR/CHF',
 'AUD/USD',
 'USD/CAD',
 'NZD/USD',
 'EUR/GBP',
 'EUR/JPY',
 'GBP/JPY',
 'CHF/JPY',
 'GBP/CHF',
 'EUR/AUD',
 'EUR/CAD',
 'AUD/CAD',
 'AUD/JPY',
 'CAD/JPY',
 'NZD/JPY',
 'GBP/CAD',
 'GBP/NZD',
 'GBP/AUD',
 'AUD/NZD',
 'USD/SEK',
 'EUR/SEK',
 'EUR/NOK',
 'USD/NOK',
 'USD/MXN',
 'AUD/CHF',
 'EUR/NZD',
 'USD/ZAR',
 'USD/HKD',
 'ZAR/JPY',
 'USD/TRY',
 'EUR/TRY',
 'NZD/CHF',
 'CAD/CHF',
 'NZD/CAD',
 'USD/HUF',
 'TRY/JPY',
 'USD/CNH',
 'AUS200',
 'ESP35',
 'FRA40',
 'GER30',
 'HKG33',
 'JPN225',
 'NAS100',
 'SPX500',
 'UK100',
 'US30',
 'Copper',
 'CHN50',
 'EUSTX50',
 'VOLX',
 'USDOLLAR',
 'US2000',
 'INDIA50',
 'USOil',
 'UKOil',
 'SOYF',
 'NGAS',
 'USOilSpot',
 'UKOilSpot',
 'WHEATF',
 'CORNF',
 'Bund',
 'XAU/USD',
 'XAG/USD',
 'EMBasket',
 'JPYBasket',
 'BTC/USD',
 'BCH/USD',
 'ETH/USD',
 'LTC/USD',
 'CryptoMajor',
 'EOS/USD',
 'XLM/USD',
 'ESPORTS',
 'BIOTECH',
 'CANNABIS',
 'FAANG',
 'CHN.TECH',
 'CHN.ECOMM',
 'USEquities',
 'A

In [11]:
len(instr)

272

There are a total of 272 instruments at fxcm for trading.

In [12]:
#to close the connection
api.close()


## Getting Historical Data (Part 1)

To write our own algorithms and automate the process of trading , historical data is very important.

In [13]:
# importing Libraries
import pandas as pd
import fxcmpy

In [14]:
#to start a connection
api = fxcmpy.fxcmpy(config_file= "fxcm.cfg")

In [16]:
#parameters of get_candles method
help(api.get_candles)

Help on method get_candles in module fxcmpy.fxcmpy:

get_candles(instrument='', offer_id=None, period='H1', number=10, start=None, end=None, with_index=True, columns=[], stop=None) method of fxcmpy.fxcmpy.fxcmpy instance
    Return historical data from the fxcm database as pandas.DataFrame.
    
    Arguments:
    
    instrument: string (default ''):
        the instrument for which data is requested. For a list of all
        available instruments for historical data, use
        get_instruments_for_candles().
        If the value is equal to '' (default), offer_id must have a value. 
        If both, instrument and offer_id are given, the value of instrument 
        is used.
    
    offer_id: integer (default None): 
        the id of the instrument for which data is requested as given in the
        offer trading table as given by get_offers(). If offer_id is equal 
        to None (default), the parameter instrument must have a value. 
        If both, instrument and offer_id ar

In [17]:
#getting candles of EUR/USD
api.get_candles('EUR/USD')
#default gives us latest 10 candles

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2021-09-10 11:00:00  1.18301   1.18330  1.18368  1.18281  1.18313   1.18341   
2021-09-10 12:00:00  1.18330   1.18263  1.18409  1.18255  1.18341   1.18274   
2021-09-10 13:00:00  1.18263   1.18343  1.18345  1.18207  1.18274   1.18355   
2021-09-10 14:00:00  1.18343   1.18326  1.18372  1.18290  1.18355   1.18338   
2021-09-10 15:00:00  1.18326   1.18211  1.18335  1.18177  1.18338   1.18223   
2021-09-10 16:00:00  1.18211   1.18179  1.18226  1.18150  1.18223   1.18191   
2021-09-10 17:00:00  1.18179   1.18147  1.18181  1.18116  1.18191   1.18159   
2021-09-10 18:00:00  1.18147   1.18148  1.18166  1.18116  1.18159   1.18159   
2021-09-10 19:00:00  1.18148   1.18122  1.18162  1.18118  1.18159   1.18133   
2021-09-10 20:00:00  1.18122   1.18078  1.18129  1.18078  1.18133   1.18136   

                     askhigh   asklow  tickqty  
date                                            
2021-09-10 11:00:00  1.18380  1.18293     7576  
2021-09-10 12:00:00  1.18428  1.18266    10258  
2021-09-10 13:00:00  1.18357  1.18219    13081  
2021-09-10 14:00:00  1.18384  1.18302    14793  
2021-09-10 15:00:00  1.18346  1.18189    12397  
2021-09-10 16:00:00  1.18237  1.18163     7577  
2021-09-10 17:00:00  1.18192  1.18128     4553  
2021-09-10 18:00:00  1.18178  1.18128     4105  
2021-09-10 19:00:00  1.18174  1.18128     4431  
2021-09-10 20:00:00  1.18142  1.18104     4757

In [19]:
#getting latest 10,000 candles
api.get_candles('EUR/USD', number = 10000)


bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2020-02-13 09:00:00  1.08864   1.08823  1.08883  1.08810  1.08875   1.08837   
2020-02-13 10:00:00  1.08823   1.08733  1.08872  1.08733  1.08837   1.08746   
2020-02-13 11:00:00  1.08733   1.08571  1.08751  1.08555  1.08746   1.08583   
2020-02-13 12:00:00  1.08571   1.08632  1.08690  1.08525  1.08583   1.08645   
2020-02-13 13:00:00  1.08632   1.08543  1.08686  1.08527  1.08645   1.08554   
...                      ...       ...      ...      ...      ...       ...   
2021-09-10 16:00:00  1.18211   1.18179  1.18226  1.18150  1.18223   1.18191   
2021-09-10 17:00:00  1.18179   1.18147  1.18181  1.18116  1.18191   1.18159   
2021-09-10 18:00:00  1.18147   1.18148  1.18166  1.18116  1.18159   1.18159   
2021-09-10 19:00:00  1.18148   1.18122  1.18162  1.18118  1.18159   1.18133   
2021-09-10 20:00:00  1.18122   1.18078  1.18129  1.18078  1.18133   1.18136   

                     askhigh   asklow  tickqty  
date                                            
2020-02-13 09:00:00  1.08894  1.08822     5715  
2020-02-13 10:00:00  1.08882  1.08745     6092  
2020-02-13 11:00:00  1.08761  1.08565     9397  
2020-02-13 12:00:00  1.08700  1.08537    10213  
2020-02-13 13:00:00  1.08697  1.08539     8292  
...                      ...      ...      ...  
2021-09-10 16:00:00  1.18237  1.18163     7577  
2021-09-10 17:00:00  1.18192  1.18128     4553  
2021-09-10 18:00:00  1.18178  1.18128     4105  
2021-09-10 19:00:00  1.18174  1.18128     4431  
2021-09-10 20:00:00  1.18142  1.18104     4757  

[10000 rows x 9 columns]

In [20]:
df = api.get_candles('EUR/USD', number = 10000)
df

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2020-02-13 09:00:00  1.08864   1.08823  1.08883  1.08810  1.08875   1.08837   
2020-02-13 10:00:00  1.08823   1.08733  1.08872  1.08733  1.08837   1.08746   
2020-02-13 11:00:00  1.08733   1.08571  1.08751  1.08555  1.08746   1.08583   
2020-02-13 12:00:00  1.08571   1.08632  1.08690  1.08525  1.08583   1.08645   
2020-02-13 13:00:00  1.08632   1.08543  1.08686  1.08527  1.08645   1.08554   
...                      ...       ...      ...      ...      ...       ...   
2021-09-10 16:00:00  1.18211   1.18179  1.18226  1.18150  1.18223   1.18191   
2021-09-10 17:00:00  1.18179   1.18147  1.18181  1.18116  1.18191   1.18159   
2021-09-10 18:00:00  1.18147   1.18148  1.18166  1.18116  1.18159   1.18159   
2021-09-10 19:00:00  1.18148   1.18122  1.18162  1.18118  1.18159   1.18133   
2021-09-10 20:00:00  1.18122   1.18078  1.18129  1.18078  1.18133   1.18136   

                     askhigh   asklow  tickqty  
date                                            
2020-02-13 09:00:00  1.08894  1.08822     5715  
2020-02-13 10:00:00  1.08882  1.08745     6092  
2020-02-13 11:00:00  1.08761  1.08565     9397  
2020-02-13 12:00:00  1.08700  1.08537    10213  
2020-02-13 13:00:00  1.08697  1.08539     8292  
...                      ...      ...      ...  
2021-09-10 16:00:00  1.18237  1.18163     7577  
2021-09-10 17:00:00  1.18192  1.18128     4553  
2021-09-10 18:00:00  1.18178  1.18128     4105  
2021-09-10 19:00:00  1.18174  1.18128     4431  
2021-09-10 20:00:00  1.18142  1.18104     4757  

[10000 rows x 9 columns]

In [21]:
df.info()

<class 'pandas.core.frame.DataFrame'>
DatetimeIndex: 10000 entries, 2020-02-13 09:00:00 to 2021-09-10 20:00:00
Data columns (total 9 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   bidopen   10000 non-null  float64
 1   bidclose  10000 non-null  float64
 2   bidhigh   10000 non-null  float64
 3   bidlow    10000 non-null  float64
 4   askopen   10000 non-null  float64
 5   askclose  10000 non-null  float64
 6   askhigh   10000 non-null  float64
 7   asklow    10000 non-null  float64
 8   tickqty   10000 non-null  int64  
dtypes: float64(8), int64(1)
memory usage: 781.2 KB


In [22]:
#we can also select particular columns for better visualisation of data.
api.get_candles('EUR/USD', number = 5, columns = ["bids"])

bidopen  bidclose  bidhigh   bidlow
date                                                    
2021-09-10 16:00:00  1.18211   1.18179  1.18226  1.18150
2021-09-10 17:00:00  1.18179   1.18147  1.18181  1.18116
2021-09-10 18:00:00  1.18147   1.18148  1.18166  1.18116
2021-09-10 19:00:00  1.18148   1.18122  1.18162  1.18118
2021-09-10 20:00:00  1.18122   1.18078  1.18129  1.18078

In [23]:
api.get_candles('EUR/USD', number = 5, columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2021-09-10 16:00:00  1.18223   1.18191  1.18237  1.18163
2021-09-10 17:00:00  1.18191   1.18159  1.18192  1.18128
2021-09-10 18:00:00  1.18159   1.18159  1.18178  1.18128
2021-09-10 19:00:00  1.18159   1.18133  1.18174  1.18128
2021-09-10 20:00:00  1.18133   1.18136  1.18142  1.18104

In [24]:
#getting candles for some more instruments. 
# Nifty50
api.get_candles('INDIA50')

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2021-09-10 11:00:00  17423.6   17424.6  17444.1  17421.6  17428.1   17425.6   
2021-09-10 12:00:00  17424.1   17405.1  17428.1  17403.6  17425.6   17407.1   
2021-09-10 13:00:00  17405.1   17409.1  17418.6  17400.6  17407.6   17411.6   
2021-09-10 14:00:00  17409.1   17378.6  17412.6  17377.1  17411.1   17380.6   
2021-09-10 15:00:00  17378.6   17370.1  17379.6  17355.1  17381.6   17373.6   
2021-09-10 16:00:00  17370.6   17383.6  17389.6  17366.1  17373.6   17386.1   
2021-09-10 17:00:00  17384.1   17385.1  17391.6  17381.1  17386.1   17388.1   
2021-09-10 18:00:00  17385.1   17375.6  17388.1  17361.6  17387.6   17377.1   
2021-09-10 19:00:00  17375.1   17342.1  17376.1  17339.6  17377.6   17344.1   
2021-09-10 20:00:00  17342.1   17338.6  17346.1  17338.6  17344.1   17342.1   

                     askhigh   asklow  tickqty  
date                                            
2021-09-10 11:00:00  17445.6  17425.1     1394  
2021-09-10 12:00:00  17429.6  17405.1     2127  
2021-09-10 13:00:00  17421.6  17402.1     4468  
2021-09-10 14:00:00  17415.1  17379.1     4345  
2021-09-10 15:00:00  17381.6  17357.6     3313  
2021-09-10 16:00:00  17390.1  17368.6     2399  
2021-09-10 17:00:00  17393.6  17382.6     2638  
2021-09-10 18:00:00  17390.6  17362.6     2092  
2021-09-10 19:00:00  17378.1  17341.6     3772  
2021-09-10 20:00:00  17348.1  17340.6      886

## Getting Historical Data (Part 2)

In [25]:
api.get_candles('EUR/USD', number = 10000, columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2020-02-13 09:00:00  1.08875   1.08837  1.08894  1.08822
2020-02-13 10:00:00  1.08837   1.08746  1.08882  1.08745
2020-02-13 11:00:00  1.08746   1.08583  1.08761  1.08565
2020-02-13 12:00:00  1.08583   1.08645  1.08700  1.08537
2020-02-13 13:00:00  1.08645   1.08554  1.08697  1.08539
...                      ...       ...      ...      ...
2021-09-10 16:00:00  1.18223   1.18191  1.18237  1.18163
2021-09-10 17:00:00  1.18191   1.18159  1.18192  1.18128
2021-09-10 18:00:00  1.18159   1.18159  1.18178  1.18128
2021-09-10 19:00:00  1.18159   1.18133  1.18174  1.18128
2021-09-10 20:00:00  1.18133   1.18136  1.18142  1.18104

[10000 rows x 4 columns]

In [26]:
# we can also specify start and end date , to get past data
api.get_candles('EUR/USD', start = "2020-07-01", end = "2020-07-31", columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2020-07-01 00:00:00  1.12341   1.12313  1.12342  1.12282
2020-07-01 01:00:00  1.12313   1.12394  1.12402  1.12296
2020-07-01 02:00:00  1.12394   1.12386  1.12414  1.12370
2020-07-01 03:00:00  1.12386   1.12271  1.12395  1.12228
2020-07-01 04:00:00  1.12271   1.12188  1.12279  1.12157
...                      ...       ...      ...      ...
2020-07-30 20:00:00  1.18452   1.18479  1.18489  1.18437
2020-07-30 21:00:00  1.18479   1.18425  1.18483  1.18379
2020-07-30 22:00:00  1.18425   1.18488  1.18493  1.18409
2020-07-30 23:00:00  1.18488   1.18728  1.18772  1.18437
2020-07-31 00:00:00  1.18728   1.18668  1.18894  1.18625

[538 rows x 4 columns]

In [27]:
# we can also specify the length of each candle , default is 1 hour (H1).
api.get_candles('EUR/USD', start = "2020-07-01", end = "2020-07-31", 
                period = "H1", columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2020-07-01 00:00:00  1.12341   1.12313  1.12342  1.12282
2020-07-01 01:00:00  1.12313   1.12394  1.12402  1.12296
2020-07-01 02:00:00  1.12394   1.12386  1.12414  1.12370
2020-07-01 03:00:00  1.12386   1.12271  1.12395  1.12228
2020-07-01 04:00:00  1.12271   1.12188  1.12279  1.12157
...                      ...       ...      ...      ...
2020-07-30 20:00:00  1.18452   1.18479  1.18489  1.18437
2020-07-30 21:00:00  1.18479   1.18425  1.18483  1.18379
2020-07-30 22:00:00  1.18425   1.18488  1.18493  1.18409
2020-07-30 23:00:00  1.18488   1.18728  1.18772  1.18437
2020-07-31 00:00:00  1.18728   1.18668  1.18894  1.18625

[538 rows x 4 columns]

In [30]:
#D1 --> 1 Day
api.get_candles('EUR/USD', start = "2020-07-01", end = "2020-07-31", 
                period = "D1", columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2020-07-02 21:00:00  1.12522   1.12404  1.13032  1.12239
2020-07-03 21:00:00  1.12404   1.12498  1.12544  1.12197
2020-07-05 21:00:00  1.12498   1.12478  1.12514  1.12462
2020-07-06 21:00:00  1.12478   1.13100  1.13461  1.12478
2020-07-07 21:00:00  1.13100   1.12748  1.13331  1.12594
2020-07-08 21:00:00  1.12748   1.13309  1.13522  1.12629
2020-07-09 21:00:00  1.13309   1.12856  1.13713  1.12808
2020-07-10 21:00:00  1.12856   1.13020  1.13254  1.12553
2020-07-12 21:00:00  1.13020   1.13079  1.13120  1.13025
2020-07-13 21:00:00  1.13079   1.13451  1.13755  1.13015
2020-07-14 21:00:00  1.13451   1.14003  1.14093  1.13258
2020-07-15 21:00:00  1.14003   1.14124  1.14527  1.13916
2020-07-16 21:00:00  1.14124   1.13843  1.14423  1.13711
2020-07-17 21:00:00  1.13843   1.14299  1.14442  1.13782
2020-07-19 21:00:00  1.14299   1.14200  1.14327  1.14037
2020-07-20 21:00:00  1.14200   1.14483  1.14683  1.14029
2020-07-21 21:00:00  1.14483   1.15276  1.15403  1.14236
2020-07-22 21:00:00  1.15276   1.15704  1.16021  1.15074
2020-07-23 21:00:00  1.15704   1.15968  1.16275  1.15409
2020-07-24 21:00:00  1.15968   1.16589  1.16591  1.15818
2020-07-26 21:00:00  1.16589   1.16456  1.16587  1.16405
2020-07-27 21:00:00  1.16456   1.17523  1.17819  1.16429
2020-07-28 21:00:00  1.17523   1.17168  1.17744  1.16993
2020-07-29 21:00:00  1.17168   1.17923  1.18069  1.17147
2020-07-30 21:00:00  1.17923   1.18479  1.18489  1.17314
2020-07-31 21:00:00  1.18479   1.17829  1.19093  1.17618

In [32]:
# m1 --> 1 minutes.
api.get_candles('EUR/USD', start = "2020-07-01", end = "2020-07-10", 
                period = "m1", columns = ["asks"])

askopen  askclose  askhigh   asklow
date                                                    
2020-07-01 00:00:00  1.12341   1.12331  1.12342  1.12330
2020-07-01 00:01:00  1.12331   1.12317  1.12332  1.12315
2020-07-01 00:02:00  1.12317   1.12310  1.12317  1.12308
2020-07-01 00:03:00  1.12310   1.12301  1.12312  1.12291
2020-07-01 00:04:00  1.12301   1.12299  1.12302  1.12293
...                      ...       ...      ...      ...
2020-07-09 23:56:00  1.12899   1.12891  1.12899  1.12890
2020-07-09 23:57:00  1.12891   1.12891  1.12891  1.12890
2020-07-09 23:58:00  1.12891   1.12884  1.12891  1.12883
2020-07-09 23:59:00  1.12884   1.12887  1.12887  1.12881
2020-07-10 00:00:00  1.12887   1.12864  1.12895  1.12863

[10002 rows x 4 columns]

In [42]:
#Example
api.get_candles("INDIA50" ,  start = "2020-10-01", end = "2020-11-10")

bidopen  bidclose  bidhigh   bidlow  askopen  askclose  \
date                                                                          
2020-11-09 04:00:00  12448.5   12432.5  12457.5  12423.0  12449.5   12433.0   
2020-11-09 05:00:00  12431.5   12411.5  12447.0  12395.0  12433.0   12413.0   
2020-11-09 06:00:00  12411.5   12403.5  12421.0  12383.5  12413.0   12404.5   
2020-11-09 07:00:00  12404.5   12427.5  12429.0  12387.5  12404.5   12428.0   
2020-11-09 08:00:00  12427.5   12438.5  12454.0  12422.0  12427.5   12439.5   
2020-11-09 09:00:00  12439.0   12485.0  12485.5  12439.0  12439.5   12485.5   
2020-11-09 10:00:00  12485.5   12493.5  12497.0  12471.5  12485.5   12494.5   
2020-11-09 11:00:00  12493.0   12601.5  12654.5  12475.5  12494.0   12605.5   
2020-11-09 12:00:00  12602.0   12766.5  12779.5  12562.5  12606.5   12768.5   
2020-11-09 13:00:00  12768.0   12755.5  12788.0  12739.5  12768.5   12757.0   
2020-11-09 14:00:00  12755.5   12680.5  12819.5  12678.0  12757.5   12682.0   
2020-11-09 15:00:00  12680.5   12702.5  12739.5  12676.5  12682.0   12703.5   
2020-11-09 16:00:00  12702.5   12726.5  12734.0  12683.5  12702.5   12728.0   
2020-11-09 17:00:00  12726.5   12777.0  12785.5  12726.5  12728.0   12778.0   
2020-11-09 18:00:00  12778.0   12756.0  12784.0  12742.0  12779.0   12757.5   
2020-11-09 19:00:00  12756.0   12756.5  12766.0  12729.5  12757.5   12757.0   
2020-11-09 20:00:00  12756.5   12678.5  12761.5  12674.0  12757.0   12681.5   
2020-11-09 21:00:00  12676.5   12679.0  12683.0  12673.0  12681.0   12688.0   
2020-11-09 22:00:00  12673.0   12673.0  12673.0  12673.0  12689.0   12689.0   

                     askhigh   asklow  tickqty  
date                                            
2020-11-09 04:00:00  12458.0  12423.5     6061  
2020-11-09 05:00:00  12447.5  12395.0     7817  
2020-11-09 06:00:00  12421.5  12385.0     6510  
2020-11-09 07:00:00  12430.5  12388.5     6595  
2020-11-09 08:00:00  12455.0  12422.5     6770  
2020-11-09 09:00:00  12486.0  12439.5     6146  
2020-11-09 10:00:00  12499.0  12483.0      700  
2020-11-09 11:00:00  12660.0  12476.5     2514  
2020-11-09 12:00:00  12781.0  12567.5     7082  
2020-11-09 13:00:00  12789.0  12740.5     3859  
2020-11-09 14:00:00  12821.0  12679.0     5173  
2020-11-09 15:00:00  12741.0  12677.5     6009  
2020-11-09 16:00:00  12735.5  12685.0     4539  
2020-11-09 17:00:00  12787.5  12728.0     3278  
2020-11-09 18:00:00  12785.0  12743.0     2032  
2020-11-09 19:00:00  12768.0  12730.5     2263  
2020-11-09 20:00:00  12763.0  12674.5     2881  
2020-11-09 21:00:00  12688.5  12677.0      267  
2020-11-09 22:00:00  12689.0  12689.0        2

In [43]:
api.close()

## Streaming high-frequency real-time Data

To forward test our trading Algorithms , real-time data is must.

In [44]:
# importing Libraries
import pandas as pd
import fxcmpy

In [45]:
# establishing a connection
api = fxcmpy.fxcmpy(config_file= "fxcm.cfg")

In [46]:
#subscribing EUR/USD
api.subscribe_market_data("EUR/USD")

In [47]:
# All subscribed instruments
api.get_subscribed_symbols()

['EUR/USD']

In [48]:
api.get_last_price("EUR/USD")

Bid     1.18078
Ask     1.18136
High    1.18518
Low     1.18078
Name: 2021-09-10 20:59:14.843000, dtype: float64

In [49]:
api.get_prices("EUR/USD")

Bid      Ask     High      Low
2021-09-10 20:59:14.843  1.18078  1.18136  1.18518  1.18078

In [50]:
import time

In [51]:
while True:
    time.sleep(1)
    print(api.get_last_price("EUR/USD").name, api.get_last_price("EUR/USD").Bid,
          api.get_last_price("EUR/USD").Ask, sep = "|")

2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136
2021-09-10 20:59:14.843000|1.18078|1.18136


KeyboardInterrupt: 

In [52]:
api.unsubscribe_market_data("EUR/USD")

In [53]:
def print_data(data, dataframe):
    print('%3d | %s | %s, %s, %s'
          % (len(dataframe), data['Symbol'],
             pd.to_datetime(int(data['Updated']), unit='ms'),
             data['Rates'][0], data['Rates'][1]))

In [54]:
api.subscribe_market_data("EUR/USD", (print_data,))

In [55]:
api.unsubscribe_market_data("EUR/USD")

In [56]:
api.close()

## Creating Orders and Executing Trades

In [57]:
import pandas as pd
import fxcmpy

In [58]:
api = fxcmpy.fxcmpy(config_file= "fxcm.cfg")

In [59]:
api.get_open_positions()

Empty DataFrame
Columns: []
Index: []

In [70]:
api.create_market_buy_order("EUR/USD", 100)

In [61]:
api.get_open_positions()

t  ratePrecision   tradeId accountName accountId  roll  com     open  \
0  1              5  71250282  1001682745   1682745     0    0  1.18136   

  valueDate  grossPL  ...  usedMargin    orderId  OpenOrderRequestTXT  \
0                -58  ...         325  132057202                 None   

   OpenOrderReqID  TradeIDOrigin  child_trailingStop  child_trailing stop  \
0  Request-232275           None                none              -1    0   

   stopMove limit  
0         0     0  

[1 rows x 28 columns]

In [62]:
col = ["tradeId", "amountK", "currency", "grossPL", "isBuy"]

In [63]:
api.get_open_positions()[col]

tradeId  amountK currency  grossPL  isBuy
0  71250282      100  EUR/USD      -58   True

In [64]:
api.create_market_sell_order("EUR/USD", 100)

In [65]:
api.get_open_positions()[col]

tradeId  amountK currency  grossPL  isBuy
0  71250282      100  EUR/USD      -58   True
1  71250283      100  EUR/USD      -58  False

In [66]:
api.close_all_for_symbol("EUR/USD")

In [67]:
api.get_closed_positions()[col]

KeyError: 'closed_positions'

In [68]:
api.get_closed_positions_summary()[col]

tradeId  amountK currency  grossPL  isBuy
0              200  EUR/USD     -116  False

In [69]:
api.get_accounts().T

0
t                           6
ratePrecision               0
accountId             1682745
balance              20065.15
usdMr                       0
mc                          N
mcDate                       
accountName        1001682745
usdMr3                      0
hedging                     Y
usableMargin3        20065.15
usableMarginPerc          100
usableMargin3Perc         100
equity               20065.15
usableMargin         20065.15
bus                      1000
dayPL                    -116
grossPL                     0

In [ ]:
api.close()